### PR TITLE
[WIP] Windows dev environment configuration, update install instructions from source in the docs

### PR DESCRIPTION
**Pull request by vexilligera**
## Description ##
Taking over #17206 with fix to #17635, update toolchains to vs2019 and cuda 10.2

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: ci/windows_dev_env/windows_deps_headless_installer.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+ssl._create_default_https_context = ssl._create_unverified_context
+
+log = logging.getLogger(__name__)
+
+
+DEPS = {
+    'openblas': 'https://windows-post-install.s3-us-west-2.amazonaws.com/OpenBLAS-windows-v0_2_19.zip',
+    'opencv': 'https://windows-post-install.s3-us-west-2.amazonaws.com/opencv-windows-4.1.2-vc14_vc15.zip',
+    'cudnn': 'https://windows-post-install.s3-us-west-2.amazonaws.com/cudnn-9.2-windows10-x64-v7.4.2.24.zip',
+    'nvdriver': 'https://windows-post-install.s3-us-west-2.amazonaws.com/nvidia_display_drivers_398.75_server2016.zip',
```
Comment (leezu) [inline ci/windows_dev_env/windows_deps_headless_installer.py]: Who owns the `windows-post-install` bucket?

File: ci/windows_dev_env/windows_deps_headless_installer.py [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+
+def install_perl():
+    logging.info("Installing Perl")
+    with tempfile.TemporaryDirectory() as tmpdir:
+        perl_file_path = download(DEPS['perl'], tmpdir)
+        check_call(['msiexec ', '/n', '/passive', '/i', perl_file_path])
+    logging.info("Perl install complete")
+
+
+def install_clang():
```
Comment (leezu) [inline ci/windows_dev_env/windows_deps_headless_installer.py]: Why do we need clang when compiling with visual studio?

File: ci/windows_dev_env/windows_deps_headless_installer.py [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+            logging.info("Waiting for Visual Studio to install for the last {} seconds".format(str(timer)))
+        sleep(1)
+        timer += 1
+    if vs_still_installing():
+        logging.warning("VS install still running after timeout (%d)", DEFAULT_SUBPROCESS_TIMEOUT)
+    else:
+        logging.info("Visual studio install complete.")
+
+
+def install_cmake():
```
Comment (leezu) [inline ci/windows_dev_env/windows_deps_headless_installer.py]: Unused

File: ci/windows_dev_env/windows_deps_headless_installer.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+        with zipfile.ZipFile(local_file, 'r') as zip:
+            zip.extractall(tmpdir)
+        with remember_cwd():
+            os.chdir(tmpdir)
+            check_call(".\setup.exe -noreboot -clean -noeula -nofinish -passive")
+    logging.info("NVidia install complete")
+
+
+def install_cuda():
+    # CUDA 9.2 and patches
```
Comment (leezu) [inline ci/windows_dev_env/windows_deps_headless_installer.py]: Enforcing Cuda 9.2 may not be very future proof for this script.

File: docs/static_site/src/pages/get_started/windows_setup.md [hunk 5]
```diff
@@ -137,126 +137,37 @@ Check the chart below for other options or refer to [PyPI for other MXNet pip pa
...
-3. Verify that the `DCUDNN_INCLUDE` and `DCUDNN_LIBRARY` environment variables are pointing to the `include` folder and `cudnn.lib` file of your CUDA installed location, and `C:\incubator-mxnet` is the location of the source code you just cloned in the previous step.
-4. Create a build dir using the following command and go to the directory, for example:
-```
-mkdir C:\incubator-mxnet\build
-cd C:\incubator-mxnet\build
-```
-5. Compile the MXNet source code with `cmake` by using following command:
 ```
-cmake -G "Visual Studio 15 2017 Win64" -T cuda=9.2,host=x64 -DUSE_CUDA=1 -DUSE_CUDNN=1 -DUSE_NVRTC=1 -DUSE_OPENCV=1 -DUSE_OPENMP=1 -DUSE_BLAS=open -DUSE_LAPACK=1 -DUSE_DIST_KVSTORE=0 -DCUDA_ARCH_LIST=Common -DCUDA_TOOLSET=9.2 -DCUDNN_INCLUDE=C:\cuda\include -DCUDNN_LIBRARY=C:\cuda\lib\x64\cudnn.lib "C:\incubator-mxnet"
+.\setup.ps1
```
Comment (leezu) [inline docs/static_site/src/pages/get_started/windows_setup.md]: Is it feasible to provide short non-blackbox instructions to install MXNet?

For example, if users have a standard installation of visual studio, can the installation experience be pretty similar to https://mxnet.apache.org/get_started/ubuntu_setup#build-the-mxnet-shared-library-from-source Step 2 and Step 3?

**Comment by aaronmarkham** [comment]
Let me know when this is ready, and I'll test it out.

**Comment by leezu** [comment]
@vexilligera what is the status of this PR?

**Comment by vexilligera** [comment]
> @vexilligera what is the status of this PR?

I'm trying to update the AMI to VS2019 and CUDA 10.2 and there's a bug with dmlc-core, probably caused by different MSVC standards. I'm working on that.

File: docs/static_site/src/pages/get_started/windows_setup.md (hunk 5 above)
Comment (vexilligera) [reply docs/static_site/src/pages/get_started/windows_setup.md]: Yes, we can explicitly use the cmd command invoked from ci/build_windows.py

File: ci/windows_dev_env/windows_deps_headless_installer.py (hunk 1 above)
Comment (vexilligera) [reply ci/windows_dev_env/windows_deps_headless_installer.py]: I currently don't have access to windows-post-install, so I opened another bucket (mxnet-windows-build). It would be easy to merge them.

File: ci/windows_dev_env/windows_deps_headless_installer.py (hunk 2 above)
Comment (vexilligera) [reply ci/windows_dev_env/windows_deps_headless_installer.py]: They probably need to support tvm op build on windows as well, which may need this.

File: ci/windows_dev_env/windows_deps_headless_installer.py (hunk 4 above)
Comment (vexilligera) [reply ci/windows_dev_env/windows_deps_headless_installer.py]: CUDA is updated to 10.2 now.

**Comment by vexilligera** [comment]
> Making it public is redistribution and thus license infringement.

I may also need to upload the VS2019 web installer as in `install_vs`, since the original download page needs redirection to the file url and I couldn't find any short url as for VS2017.

**Comment by ChaiBapchya** [comment]
Updated links for cudnn & VS2019
- https://windows-post-install.s3-us-west-2.amazonaws.com/vs_community__1246179388.1585201415.exe
- https://windows-post-install.s3-us-west-2.amazonaws.com/cudnn-10.2-windows10-x64-v7.6.5.32.zip

File: ci/windows_dev_env/windows_deps_headless_installer.py (hunk 1 above)
Comment (larroy) [reply ci/windows_dev_env/windows_deps_headless_installer.py]: mxnet-ci-dev

**Comment by leezu** [comment]
@mxnet-bot run ci [windows-cpu, windows-gpu]

**Comment by ChaiBapchya** [comment]
retriggering.. windows

**Comment by ChaiBapchya** [comment]
It's failing only for windows-cpu MKLDNN MKL part with the error
"can't find specified path"
```
[2020-03-27T18:52:40.264Z] "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvarsx86_amd64.bat" && cmake -G Ninja -DCMAKE_C_COMPILER=cl -DCMAKE_CXX_COMPILER=cl -DUSE_CUDA=OFF 
[2020-03-27T18:52:40.264Z] -DUSE_CUDNN=OFF -DENABLE_CUDA_RTC=OFF -DUSE_OPENCV=ON -DOpenCV_RUNTIME=vc15 -DOpenCV_ARCH=x64 -DUSE_OPENMP=ON -DUSE_BLAS=mkl -DUSE_LAPACK=ON -DUSE_DIST_KVSTORE=OFF -DUSE_MKL_IF_AVAILABLE=ON 
[2020-03-27T18:52:40.264Z] -DUSE_MKLDNN=ON -DCMAKE_BUILD_TYPE=Release C:\jenkins_slave\workspace\build-cpu-mkldnn-mkl
```

@larroy is this related to path not being on oneline? or the path itself is incorrect?

**Comment by ChaiBapchya** [comment]
Turns out the path mentioned in ci/build_windows.py is correct
```
VS 2019': r'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvarsx86_amd64.bat
```
That is exactly where the bat file is.

[verified this by launching an instance using the updated windows AMI]

**Comment by vexilligera** [comment]
WIN_GPU_MKLDNN is having an issue similar to https://github.com/pytorch/pytorch/issues/25393

**Comment by vexilligera** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by vexilligera** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by vexilligera** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by ChaiBapchya** [comment]
@marcoabreu Even in this case, @vexilligera is unable to reproduce the error locally. And Windows GPU + MKLDNN is flaky. How do you suggest we proceed trying to resolve the thrust issue?

Fail : http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fwindows-gpu/detail/PR-17808/13/pipeline
Pass : http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fwindows-gpu/detail/PR-17808/12/pipeline

**Comment by ChaiBapchya** [comment]
@vexilligera as discussed offline, lets try testing locally for WIN_GPU and WIN_GPU_MKLDNN build 10 times each (since 1 run takes 20-30mins) to come up with some basis... (ideally would have tried 100 times but given the resource & time constraints)

**Comment by vexilligera** [comment]
> @vexilligera as discussed offline, lets try testing locally for WIN_GPU and WIN_GPU_MKLDNN build 10 times each (since 1 run takes 20-30mins) to come up with some basis... (ideally would have tried 100 times but given the resource & time constraints)

On my local test, the WIN_GPU_MKLDNN is much more flaky than WIN_GPU, as all WIN_GPU builds passed while about 1/3 of WIN_GPU_MKLDNN builds failed.

@haojin2 suggests us removing the WIN_GPU_MKLDNN test entirely since MKLDNN doesn't make much sense as we are running on GPU, and GPU_MKLDNN case is covered on other platforms.

**Comment by ChaiBapchya** [comment]
We can get rid of WIN_GPU_MKLDNN tests altogether but that still leaves us with the flakiness of WIN_GPU as can be seen in these builds
For roughly same code of this PR & same windows AMI, below are the results so far

WIN_GPU | WIN_GPU_MKLDNN | Build Number | Link
-- | -- | -- | -- |
✖︎| ✔︎|15 | http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fwindows-gpu/detail/PR-17808/15/pipeline
✖︎| ✔︎| 14 | http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fwindows-gpu/detail/PR-17808/14/pipeline
✔︎|✔︎| 12 | http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fwindows-gpu/detail/PR-17808/12/pipeline
✔︎| ✖︎| 13 | http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fwindows-gpu/detail/PR-17808/13/pipeline

Ofcourse your tests on local have a different story to tell...

**Comment by vexilligera** [comment]
@larroy do you have any idea on how to solve Win 126 on CI? Basically the binaries aren't in the search path and dlopen fails due to that

**Comment by marcoabreu** [comment]
I'll have a look at the logs when I'm at my computer 

Just to be clear about two things:
1. Retries are fine for now, but must be removed before merging.
2. It is not an option to remove GPU mkldnn. It does make sense for cases where an operator is not supported by the GPU or the user wants to run some parts on CPU. Just because it breaks, it doesn't mean that we can remove it to ease our situation.

**Comment by vexilligera** [comment]
> Sorry but I don't believe that you are testing the development on a VM when this kind of commit is necessary.

Was testing on another instance and made commits on another...manually copied the code and missed some parts...

**Comment by ChaiBapchya** [comment]
>     1. Retries are fine for now, but must be removed before merging.

With the CI blocked on windows-gpu for last 4 days and this PR trying to unblock the CI with an updated AMI, I guess it would be a good way to unblock the CI by allowing retries for now.
I have created issue : https://github.com/apache/incubator-mxnet/issues/17935 for tracking this problem. 

But for unblocking CI do you think it makes sense to add the retries as a temporary hotfix?
@vexilligera tested the retries 10 times to see that adding retries safely ensures successful compilation all the time.

**Comment by vexilligera** [comment]
I printed the path to the missing dll and it was:
C:\jenkins_slave\workspace\ut-python-gpu\windows_package\lib\libmxnet.dll
However, the build log shows this is packed into windows_package as you can find here:
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fwindows-gpu/detail/PR-17808/19/

@larroy suggests packing the opencv binary as well but since opencv is already in the environment path of the AMI I guess the real problem should be something else (if opencv is missing and libmxnet.dll present you'll get segmentation fault instead, WinError 126 is triggered from Python due to wrong path)

@marcoabreu do you have any suggestion on this?

**Comment by ChaiBapchya** [comment]
OSError: [WinError 126] The specified module could not be found

This one means the libmxnet.dll is not in the DLL Search path [acc to Stackoverflow]
Fix would be to add it into the environment
https://stackoverflow.com/questions/43987081/openslide-python-import-error

What's surprising is : libmxnet.dll & mxnet_70.dll are being packed into windows_package.7z and unpacked during test phase correctly.
While WIN_CPU tests don't give this error, WIN_GPU fails

**Comment by ChaiBapchya** [comment]
@marcoabreu Also, after adding the retries, from build #18 to #22 all 5 have passed on WIN_GPU and WIN_GPU_MKLDNN compilation stage. That's a compeling reason for using retries to temporarily resolve the intermittent failure I feel.

**Comment by vexilligera** [comment]
> Test in WIN_GPU fail at
> 
> ```
> OSError: [WinError 126] The specified module could not be found
> ```
> 
> This one means the libmxnet.dll is not in the DLL Search path [acc to Stackoverflow]
> Fix would be to add it into the environment
> https://stackoverflow.com/questions/43987081/openslide-python-import-error
> 
> What's surprising is : libmxnet.dll & mxnet_70.dll are being packed into windows_package.7z and unpacked during test phase correctly.
> While WIN_CPU tests don't give this error, WIN_GPU fails

@larroy says he used to run into the problem and it turned out that it was solved by packing opencv binary. In this case the opencv is already in the environment path and it should be found. Although WinError 126 could also occur if a dependent dll is missing, in my local tests when I deliberately created that situation, instead, segmentation fault (access violation at 0x0000000000000000) is thrown.

**Comment by leezu** [comment]
> The only dependent dll of libmxnet.dll is cudart64_102.dll (not including the win32 dlls). For mxnet_70.dll, the libraries of opencv and openblas are already in the ami's environment path.

There should be more dependent dlls? For example, if built with fusion support, the cuda driver should also be a dependent library? https://github.com/apache/incubator-mxnet/issues/17858

https://github.com/apache/incubator-mxnet/blob/56e79853ad5cf98baf84454eb595c7658bef6ee6/CMakeLists.txt#L613-L621

**Comment by vexilligera** [comment]
> > The only dependent dll of libmxnet.dll is cudart64_102.dll (not including the win32 dlls). For mxnet_70.dll, the libraries of opencv and openblas are already in the ami's environment path.
> 
> There should be more dependent dlls? For example, if built with fusion support, the cuda driver should also be a dependent library? #17858
> 
> https://github.com/apache/incubator-mxnet/blob/56e79853ad5cf98baf84454eb595c7658bef6ee6/CMakeLists.txt#L613-L621

They have fixed WinError 126 and now we are looking at a segfault

**Comment by vexilligera** [comment]
@leezu Thank you for checking the instance type and changing the cuda arch to 5.2, the ci is running now. @ChaiBapchya previously the segfault was caused by specifying 7.0 on the g3 instance.

**Comment by vexilligera** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by vexilligera** [comment]
@ChaiBapchya I added instance type auto-detection and found out c5 was running the GPU ci, which caused the segfault.

File: ci/build_windows.py [hunk 6]
```diff
@@ -33,16 +33,43 @@
...
+    # https://gitlab.kitware.com/cmake/cmake/issues/18920
     'VS 2015': r'C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\x86_amd64\vcvarsx86_amd64.bat',
-    'VS 2017': r'C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvarsx86_amd64.bat'
+    'VS 2017': r'C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvars64.bat',
+    'VS 2019': r'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat'
 }
 
 
+def get_cuda_arch(default='"5.2"'):
+    # cuda arch auto detection in cmake doesn't work on windows ci
```
Comment (leezu) [inline ci/build_windows.py]: The autodetection works fine on Windows. But we are compiling on a CPU instance. Can you remove this function and specify the MXNET_CUDA_ARCH argument as is done before? This function adds unnecessary complexity.

Comment (vexilligera) [reply ci/build_windows.py]: Mistakenly added due to misconception, already removed. Sorry about this.

Comment (leezu) [reply ci/build_windows.py]: No worries. Thanks for your work on this.

**Comment by marcoabreu** [comment]
A missing dll might not mean that the dll itself is missing but that I can not be loaded to a miss in the dependency chain. I'd recommend to get the artifact and inspect it with some resources manager (similar to ldd). That should tell you which dependencies are expected and you can verify that they are available. 

Maybe also starting with a debugger could help to see a more detailed exception which explains which dependency of the dll is missing.

File: ci/build_windows.py [hunk 7]
```diff
@@ -193,14 +238,19 @@ def windows_package(args):
         for dll in dlls:
             logging.info("packing dll: %s", dll)
             shutil.copy(dll, pkgdir_lib)
+
+        if pkgdir_lib.find('gpu') != -1:
+            add_path("C:\\jenkins_slave\\workspace\\ut-python-gpu\\windows_package\\lib")
```
Comment (josephevans) [inline ci/build_windows.py:243]: Updating the path should be done in the testing phase - in ci/windows/test_py3_gpu.ps1 so the tests can find all the DLLs.

Comment (vexilligera) [reply ci/build_windows.py:243]: That's right, forgot to remove that. Will add that to the testing phase once we got the log from ad95025

**Comment by vexilligera** [comment]
@leezu seems like adding both 5.2 and 7.0 and then drop 7.0 would trigger a normal run...

File: ci/windows/test_py3_gpu.ps1 [hunk 8]
```diff
@@ -23,7 +23,20 @@ $env:MXNET_STORAGE_FALLBACK_LOG_VERBOSE=0
 $env:MXNET_SUBGRAPH_VERBOSE=0
 $env:MXNET_HOME=[io.path]::combine($PSScriptRoot, 'mxnet_home')
```
Comment (ChaiBapchya) [inline ci/windows/test_py3_gpu.ps1:25]: let's print all ENV Vars to compare state of env vars in CI run vs manual powershell test
```Get-ChildItem Env:```

**Comment by vexilligera** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by vexilligera** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by leezu** [comment]
Connectivity errors appear fixed now by switching to g4 (with faster cpu and network).

Removed the debug statements and rebased on latest master.

**Comment by leezu** [comment]
@marcoabreu Do you still stand by your veto?

> "Retries are fine for now, but must be removed before merging."

I suggest to go ahead and merge this with the retries. We don't have control over the thrust + VS code incompatibility and may need to wait for a fix on nvidia's or microsoft's side. Doesn't seem worth to block all MXNet development until the bug in "thrust + VS code" is fixed, given it's an open issue affecting other projects as well: https://github.com/pytorch/pytorch/issues/25393

File: CMakeLists.txt [hunk 9]
```diff
@@ -163,6 +163,8 @@ if(MSVC)
   add_definitions(-DDMLC_STRICT_CXX11)
   add_definitions(-DNOMINMAX)
   set(CMAKE_C_FLAGS "/MP")
+  # report an accurate value for recent C++ language standards support
+  set(CMAKE_CXX_FLAGS "/Zc:__cplusplus")
```
Comment (leezu) [inline CMakeLists.txt:167]: @ChaiBapchya @vexilligera you added this line, but it doesn't have any effect as it is overwritte in the line below

Comment (ChaiBapchya) [reply CMakeLists.txt:167]: set(flag ${flag} "secondval") does append right?
I thought we are appending to CMAKE_CXX_FLAGS

Comment (ChaiBapchya) [reply CMakeLists.txt:167]: the typo!!!! `CMAKE_CXX_FLAGS` instead of `CMAKE_C_FLAGS`

File: ci/build_windows.py [hunk 10]
```diff
@@ -106,29 +129,37 @@ class BuildFlavour(Enum):
...
         '-DENABLE_CUDA_RTC=ON '
         '-DUSE_OPENCV=ON  '
+        '-DOpenCV_RUNTIME=vc15 '
+        '-DOpenCV_ARCH=x64 '
         '-DUSE_OPENMP=ON '
         '-DUSE_BLAS=open '
         '-DUSE_LAPACK=ON '
         '-DUSE_DIST_KVSTORE=OFF '
-        '-DMXNET_CUDA_ARCH="5.2" '
+        '-DMXNET_CUDA_ARCH="7.5" '
```
Comment (marcoabreu) [inline ci/build_windows.py:144]: This change warrants further discussion. Please change it back.

File: ci/build_windows.py [hunk 11]
```diff
@@ -253,12 +284,8 @@ def main():
     system = platform.system()
     if system == 'Windows':
         logging.info("Detected Windows platform")
-        if 'OpenBLAS_HOME' not in os.environ:
-            os.environ["OpenBLAS_HOME"] = "C:\\Program Files\\OpenBLAS-v0.2.19"
-        if 'OpenCV_DIR' not in os.environ:
-            os.environ["OpenCV_DIR"] = "C:\\Program Files\\OpenCV-v3.4.1\\build"
         if 'CUDA_PATH' not in os.environ:
-            os.environ["CUDA_PATH"] = "C:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v9.2"
+            os.environ["CUDA_PATH"] = "C:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v10.2"
```
Comment (marcoabreu) [inline ci/build_windows.py:288]: Same here

File: ci/safe_docker_run.py [hunk 12]
```diff
@@ -38,8 +38,9 @@
 
 from util import config_logging
 
-DOCKER_STOP_TIMEOUT_SECONDS = 3
+DOCKER_STOP_TIMEOUT_SECONDS = 10
```
Comment (marcoabreu) [inline ci/safe_docker_run.py:41]: Unrelated change

File: ci/windows_dev_env/windows_deps_headless_installer.py [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+    local_file = download(DEPS['openblas'])
+    with zipfile.ZipFile(local_file, 'r') as zip:
+        zip.extractall("C:\\Program Files")
+    run_command("PowerShell Set-ItemProperty -path 'hklm:\\system\\currentcontrolset\\control\\session manager\\environment' -Name OpenBLAS_HOME -Value 'C:\\Program Files\\OpenBLAS-windows-v0_2_19'")
+    logging.info("Openblas Install complete")
+
+
+def install_mkl():
+    logging.info("Installing MKL 2019.3.203...")
+    file_path = download("http://registrationcenter-download.intel.com/akdlm/irc_nas/tec/15247/w_mkl_2019.3.203.exe")
```
Comment (marcoabreu) [inline ci/windows_dev_env/windows_deps_headless_installer.py:280]: Move links and similar things into constants

**Comment by marcoabreu** [comment]
Connectivity errors were never an issue with Windows slaves. It sounds like you've changed something as part of generating the AMIs. Thus, I'm hesitant to just accept a bunch of more changes (which have not been made transparent) as a result to fix a previously created issue.

Please just revert stuff back to what it was and stop putting duckttape over things. The build used to work and the slaves used to be stable. But I have only seen attempts to change things but no rootcausing (I know about the 2GB limit). Some action resulted in our build breaking and the only sane choice is to revert whatever was done and then do these efforts again.

I will not accept multiple layers of fixes which could have been avoided if the changes were properly developed and tested in a separate environment. At the moment, it seems like open heart surgery on a productive system.

File: ci/build_windows.py (hunk 10 above)
Comment (leezu) [reply ci/build_windows.py:144]: This change sets the tested cuda architecture. The CI server now uses T4 GPUs which use 7.5 arch. Thus no, I can't change this back.

File: ci/build_windows.py (hunk 11 above)
Comment (leezu) [reply ci/build_windows.py:288]: What is your concern?

File: ci/build_windows.py (hunk 10 above)
Comment (marcoabreu) [reply ci/build_windows.py:144]: Well where did we agree that we are dropping g3 instances? How is such a change in the scope of this PR?

File: ci/build_windows.py (hunk 11 above)
Comment (marcoabreu) [reply ci/build_windows.py:288]: We didn't agree on increasing the minimum cuda version

**Comment by leezu** [comment]
Discussion continued at https://github.com/apache/incubator-mxnet/pull/17962#issuecomment-608066411

File: CMakeLists.txt (hunk 9 above)
Comment (vexilligera) [reply CMakeLists.txt]: My bad. I thought it would append...

File: ci/build_windows.py (hunk 11 above)
Comment (leezu) [reply ci/build_windows.py:288]: We agreed on using VS 2019. Cuda 10 is a requirement for using VS 2019

File: ci/build_windows.py (hunk 10 above)
Comment (leezu) [reply ci/build_windows.py:144]: Let's keep using g3 instances for now then.

File: ci/build_windows.py (hunk 11 above)
Comment (marcoabreu) [reply ci/build_windows.py:288]: Are you sure about that? Usually visual studio can be upgraded by installing the vc++ toolset which then grants the possibility to compile other cuda versions. I'm not aware that this categorically excluded.

Comment (leezu) [reply ci/build_windows.py:288]: https://superuser.com/questions/1506044/installing-cuda-9-2-with-vs-2019

Comment (leezu) [reply ci/build_windows.py:288]: https://devblogs.microsoft.com/cppblog/cuda-10-1-available-now-with-support-for-latest-microsoft-visual-studio-2019-versions/

**Comment by vexilligera** [comment]
Closing since this is now fixed in #17962

**Comment by leezu** [comment]
@vexilligera it's not fixed. We need a script to generate the working AMI. https://github.com/apache/incubator-mxnet/pull/17962 is only a hotfix.

Will you be working on the automated setup or is someone else taking over the work? Please clarify. Thank you
